soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -102,18 +109,29 @@ int intel_adsp_ipc_init(const struct device *dev)
 	config->regs->tda = INTEL_ADSP_IPC_DONE;
 #endif
 	config->regs->ctl |= (INTEL_ADSP_IPC_CTL_IDIE | INTEL_ADSP_IPC_CTL_TBIE);
+
+#ifdef CONFIG_PM_DEVICE
+	pm_device_busy_clear(dev);
+#endif /* CONFIG_PM_DEVICE */
```

Comment:
Why this busy_set/clear here ? I don't get the point, this function is not doing anything that would block and possibly let the idle thread runs ... Am I missing something ?

Reply:
Those `pm_device_busy_clear`/`pm_device_busy_set` indicate whether Device is busy or not.
Sorry for copy paste of previous answer, I'll just explain what I mean:
> From the [Device Power Manager - Busy indication](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#busy-status-indication):
> 
> > (...) Suspending a device which is in the middle of a hardware transaction, such as writing to a flash memory, may lead to undefined behavior or inconsistent states. This API guards such transactions by indicating to the kernel that the device is in the middle of an operation and should not be suspended.

When we start writing IPC registers, this is a part of "hardware transaction" between DSP and Host.
We do this during sending IPC message and acknowledging it. From MTL SoC design, we assume that there may be no running transactions between Host and DSP, because we enter D3 state during idle.
However, existing IPC messaging scheduling works using [k_work_schedule](https://docs.zephyrproject.org/3.1.0/kernel/services/threads/workqueue.html#c.k_work_schedule). We put IPC tasks in SOF on the queue and those tasks are scheduled for execution in some later time.
Theoretically, it is possible to enter idle thread and cause power transition while there are still IPC messages in the queue (that were scheduled for later execution). So in practice it may happen, that we are performing a power transition while IPC message is being send and registers are written. Those busy indicators are safeguards for this case.
If the very unlikely scenario mentioned above would happen, thanks to those safeguards a call to  `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV, PM_DEVICE_ACTION_SUSPEND);` would result in a Power Manager ignoring this device and not performing transition.
I've just noticed that this is valid only if devices are put to sleep using `pm_suspend_devices` function.
Direct call to the `pm_device_action_run()` ignores Device busy state anyway. I'll add some checks on the SOF side to platform pm_notifiers.